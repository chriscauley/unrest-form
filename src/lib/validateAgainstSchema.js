import { Validator } from 'jsonschema'

export const deepRemoveEmpty = obj => {
  const new_object = {}

  Object.entries(obj).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') {
      return
    }
    if (value.constructor !== Object) {
      new_object[key] = value
      return
    }

    const result = deepRemoveEmpty(value)
    if (Object.keys(result).length > 0) {
      new_object[key] = result
    }
  })

  return new_object
}

export default (instance, schema) => {
  const v = new Validator()
  const result = v.validate(deepRemoveEmpty(instance), schema)
  if (!result.errors.length) {
    return null
  }
  const errors = {}
  result.errors.forEach(({ path, argument, message, name }) => {
    path = [...path, argument]
    if (name === 'required') {
      errors[path.join('.')] = 'This field is required.'
    } else {
      errors[path.join('.')] = message
    }
  })
  return errors
}
