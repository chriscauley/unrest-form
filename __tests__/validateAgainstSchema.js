import validateAgainstSchema, { deepRemoveEmpty } from '@/lib/validateAgainstSchema'

test('deepRemoveEmpty', () => {
  const object = {
    a: {
      b: 1,
      c: {
        null: null,
        int: 0,
        str: '',
        a: 1,
        d: {},
        e: {
          f: {}
        }
      }
    },
    b: {}
  }
  expect(deepRemoveEmpty(object)).toMatchSnapshot()
})
