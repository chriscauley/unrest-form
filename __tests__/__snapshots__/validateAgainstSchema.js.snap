// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepRemoveEmpty 1`] = `
Object {
  "a": Object {
    "b": 1,
    "c": Object {
      "a": 1,
      "int": 0,
      "str": "",
    },
  },
}
`;
